File: ios/HMP/HMP/Features/ViewModel/PlaylistViewModel.swift
import Foundation
import shared

@Observable
class PlaylistViewModel {
    // Label lists (name strings)
    var genrePlaylistName: [String] = []
    var moodPlaylistName: [String] = []
    var scenarioPlaylistName: [String] = []
    var languagePlaylistName: [String] = []
    var eraPlaylistName: [String] = []

    // User playlists
    var userCustomPlaylists: [Playlist_] = []
    var userPlaylistsState: UiState<[Playlist_]> = .idle

    // Selected playlist
    var selectedPlaylistMusic: [MusicInfo_] = []
    var selectedPlaylistState: UiState<[MusicInfo_]> = .idle

    // Artist / Album
    var selectedArtistMusic: [MusicInfo_] = []
    var selectedArtistMusicState: UiState<[MusicInfo_]> = .idle
    var selectedAlbumMusic: [MusicInfo_] = []
    var selectedAlbumMusicState: UiState<[MusicInfo_]> = .idle

    private let managePlaylistUseCase: ManagePlaylistUseCase
    private let musicLabelUseCase: MusicLabelUseCase
    private let getAllMusicUseCase: GetAllMusicUseCase
    private let settingsRepository: SettingsRepository

    init() {
        self.managePlaylistUseCase = KoinHelperKt.getManagePlaylistUseCase()
        self.musicLabelUseCase = KoinHelperKt.getMusicLabelUseCase()
        self.getAllMusicUseCase = KoinHelperKt.getGetAllMusicUseCase()
        self.settingsRepository = KoinHelperKt.getSettingsRepository()
    }

    // MARK: - System Playlist IDs

    func getCurrentPlaylistId() async -> Int64? {
        guard let value = try? await settingsRepository.getCurrentPlaylistId() else { return nil }
        return value.int64Value
    }

    func getLikedPlaylistId() async -> Int64? {
        guard let value = try? await settingsRepository.getLikedPlaylistId() else { return nil }
        return value.int64Value
    }

    func getRecentPlaylistId() async -> Int64? {
        guard let value = try? await settingsRepository.getRecentPlaylistId() else { return nil }
        return value.int64Value
    }

    // MARK: - Labels

    func getLabelName(_ name: String) -> LabelName_? {
        LabelName_.companion.match(value: name)
    }

    func loadLabels() {
        loadLabelType(.genre, into: \.genrePlaylistName)
        loadLabelType(.mood, into: \.moodPlaylistName)
        loadLabelType(.scenario, into: \.scenarioPlaylistName)
        loadLabelType(.language, into: \.languagePlaylistName)
        loadLabelType(.era, into: \.eraPlaylistName)
    }

    private func loadLabelType(_ category: LabelCategory_, into keyPath: ReferenceWritableKeyPath<PlaylistViewModel, [String]>) {
        Task {
            do {
                let labels = try await KoinHelperKt.getLabelNamesByTypeFirst(category: category)
                await MainActor.run { self[keyPath: keyPath] = labels.map { $0.name } }
            } catch {
                print("[PlaylistVM] loadLabelType failed: \(error)")
            }
        }
    }

    // MARK: - User Playlists

    func loadUserCustomPlaylists() {
        userPlaylistsState = .loading
        Task {
            do {
                let playlists = try await managePlaylistUseCase.getAllPlaylists()
                await MainActor.run {
                    self.userCustomPlaylists = playlists
                    self.userPlaylistsState = .success(playlists)
                }
            } catch {
                await MainActor.run { self.userPlaylistsState = .error(error.localizedDescription) }
            }
        }
    }

    func createPlaylist(name: String, onCreated: ((Int64) -> Void)? = nil) {
        Task {
            do {
                let id = try await managePlaylistUseCase.createPlaylist(name: name)
                await MainActor.run { onCreated?(id.int64Value) }
                loadUserCustomPlaylists()
            } catch {
                print("[PlaylistVM] createPlaylist failed: \(error)")
            }
        }
    }

    func deletePlaylist(id: Int64) {
        Task {
            do {
                try await managePlaylistUseCase.removePlaylistById(id: id)
                loadUserCustomPlaylists()
            } catch {
                print("[PlaylistVM] deletePlaylist failed: \(error)")
            }
        }
    }

    func renamePlaylist(id: Int64, newName: String) {
        Task {
            do {
                try await managePlaylistUseCase.renamePlaylist(id: id, newName: newName)
                loadUserCustomPlaylists()
            } catch {
                print("[PlaylistVM] renamePlaylist failed: \(error)")
            }
        }
    }

    func setPlaylistPinned(id: Int64, isPinned: Bool) {
        Task {
            try? await managePlaylistUseCase.setPlaylistPinned(id: id, isPinned: isPinned)
            loadUserCustomPlaylists()
        }
    }

    // MARK: - Selected Playlist

    func loadPlaylistById(_ playlistId: Int64) {
        selectedPlaylistState = .loading
        Task {
            do {
                let music = try await managePlaylistUseCase.getPlaylistById(playlistId: playlistId)
                await MainActor.run {
                    self.selectedPlaylistMusic = music
                    self.selectedPlaylistState = music.isEmpty ? .empty : .success(music)
                }
            } catch {
                await MainActor.run { self.selectedPlaylistState = .error(error.localizedDescription) }
            }
        }
    }

    func loadPlaylistByLabel(_ labelName: LabelName_) {
        selectedPlaylistState = .loading
        Task {
            do {
                let music = try await musicLabelUseCase.getMusicListByLabel(labelName: labelName)
                await MainActor.run {
                    self.selectedPlaylistMusic = music
                    self.selectedPlaylistState = music.isEmpty ? .empty : .success(music)
                }
            } catch {
                await MainActor.run { self.selectedPlaylistState = .error(error.localizedDescription) }
            }
        }
    }

    func addItemToPlaylist(playlistId: Int64, musicId: Int64, musicPath: String) {
        Task {
            try? await managePlaylistUseCase.addToPlaylist(playlistId: playlistId, musicId: musicId, musicPath: musicPath)
        }
    }

    func removeItemFromPlaylist(musicId: Int64, playlistId: Int64) {
        Task {
            try? await managePlaylistUseCase.removeItemFromPlaylist(musicId: musicId, playlistId: playlistId)
            loadPlaylistById(playlistId)
        }
    }

    // MARK: - Playlist Meta

    var playlistMeta: Playlist_? = nil

    func getPlaylistMeta(id: Int64) async -> Playlist_? {
        try? await managePlaylistUseCase.getPlaylistMeta(id: id)
    }

    func loadPlaylistMeta(id: Int64) {
        Task {
            let meta = await getPlaylistMeta(id: id)
            await MainActor.run { self.playlistMeta = meta }
        }
    }

    // MARK: - Batch Operations

    func reorderPlaylistItems(playlistId: Int64, orderedMusicIds: [Int64]) {
        Task {
            try? await managePlaylistUseCase.reorderPlaylistItems(playlistId: playlistId, orderedMusicIds: orderedMusicIds.map { KotlinLong(value: $0) })
            loadPlaylistById(playlistId)
        }
    }

    func addItemsToPlaylist(playlistId: Int64, items: [(Int64, String)]) {
        Task {
            for (musicId, musicPath) in items {
                try? await managePlaylistUseCase.addToPlaylist(playlistId: playlistId, musicId: musicId, musicPath: musicPath)
            }
            loadPlaylistById(playlistId)
        }
    }

    func updatePlaylistDescription(id: Int64, description: String?) {
        Task {
            try? await managePlaylistUseCase.updatePlaylistDescription(id: id, description: description)
            loadPlaylistMeta(id: id)
        }
    }

    func recordPlaylistPlay(playlistId: Int64) {
        Task {
            try? await managePlaylistUseCase.incrementPlaylistPlayCount(id: playlistId)
            try? await managePlaylistUseCase.setPlaylistLastPlayedAt(id: playlistId, timestamp: Int64(Date().timeIntervalSince1970 * 1000))
        }
    }

    func loadAllMusicForAddPicker() async -> [MusicInfo_] {
        (try? await getAllMusicUseCase.invoke(orderBy: "title", orderType: "ASC")) ?? []
    }

    // MARK: - Artist / Album

    func loadArtistMusicList(artistName: String) {
        selectedArtistMusicState = .loading
        Task {
            do {
                let music = try await getAllMusicUseCase.getMusicListByArtist(artistName: artistName)
                await MainActor.run {
                    self.selectedArtistMusic = music
                    self.selectedArtistMusicState = music.isEmpty ? .empty : .success(music)
                }
            } catch {
                await MainActor.run { self.selectedArtistMusicState = .error(error.localizedDescription) }
            }
        }
    }

    func loadAlbumMusicList(albumName: String) {
        selectedAlbumMusicState = .loading
        Task {
            do {
                let music = try await getAllMusicUseCase.getMusicListByAlbum(albumName: albumName)
                await MainActor.run {
                    self.selectedAlbumMusic = music
                    self.selectedAlbumMusicState = music.isEmpty ? .empty : .success(music)
                }
            } catch {
                await MainActor.run { self.selectedAlbumMusicState = .error(error.localizedDescription) }
            }
        }
    }
}

File: android/feature-ui/src/main/java/com/hearablemusic/player/ui/playlist/viewmodel/PlaylistViewModel.kt
package com.hearablemusic.player.ui.playlist.viewmodel

import android.app.Application

import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hmp.domain.music.MusicInfo
import com.hmp.domain.enum.LabelCategory
import com.hmp.domain.enum.LabelName
import com.hmp.domain.music.usecase.GetAllMusicUseCase
import com.hmp.domain.music.usecase.MusicLabelUseCase
import com.hmp.domain.playlist.Playlist
import com.hmp.domain.playlist.usecase.ManagePlaylistUseCase
import com.hmp.domain.setting.SettingsRepository
import com.hearablemusic.player.ui.common.util.UiState
import com.hearablemusic.player.ui.R
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

import kotlinx.coroutines.Job

data class PlaylistScreenUiState(
    val playlistName: String = "",
    val playlist: UiState<List<MusicInfo>> = UiState.Idle,
    val playlistMeta: Playlist? = null,
    val selectedPlaylistId: Long? = null,
    val isCustomPlaylist: Boolean = false
)

class PlaylistViewModel(
    private val application: Application,
    private val managePlaylistUseCase: ManagePlaylistUseCase,
    private val musicLabelUseCase: MusicLabelUseCase,
    private val settingsRepository: SettingsRepository,
    private val getAllMusicUseCase: GetAllMusicUseCase
) : AndroidViewModel(application) {

    val genrePlaylistName = musicLabelUseCase.getLabelNamesByType(LabelCategory.GENRE)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val moodPlaylistName = musicLabelUseCase.getLabelNamesByType(LabelCategory.MOOD)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val scenarioPlaylistName = musicLabelUseCase.getLabelNamesByType(LabelCategory.SCENARIO)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val languagePlaylistName = musicLabelUseCase.getLabelNamesByType(LabelCategory.LANGUAGE)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val eraPlaylistName = musicLabelUseCase.getLabelNamesByType(LabelCategory.ERA)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _selectedPlaylistName = MutableStateFlow("")
    val selectedPlaylistName: StateFlow<String> = _selectedPlaylistName

    private val _selectedPlaylistId = MutableStateFlow<Long?>(null)
    val selectedPlaylistId: StateFlow<Long?> = _selectedPlaylistId.asStateFlow()

    private val _selectedPlaylistMeta = MutableStateFlow<Playlist?>(null)
    val selectedPlaylistMeta: StateFlow<Playlist?> = _selectedPlaylistMeta.asStateFlow()

    private val _isCustomPlaylist = MutableStateFlow(false)
    val isCustomPlaylist: StateFlow<Boolean> = _isCustomPlaylist.asStateFlow()

    private val _selectedPlaylistState = MutableStateFlow<UiState<List<MusicInfo>>>(UiState.Idle)
    val selectedPlaylistState: StateFlow<UiState<List<MusicInfo>>> = _selectedPlaylistState

    val playlistUiState: StateFlow<PlaylistScreenUiState> = combine(
        _selectedPlaylistName,
        _selectedPlaylistState,
        _selectedPlaylistMeta,
        _selectedPlaylistId,
        _isCustomPlaylist
    ) { name, playlistState, meta, id, isCustom ->
        PlaylistScreenUiState(
            playlistName = name,
            playlist = playlistState,
            playlistMeta = meta,
            selectedPlaylistId = id,
            isCustomPlaylist = isCustom
        )
    }.stateIn(
        viewModelScope,
        SharingStarted.WhileSubscribed(5000),
        PlaylistScreenUiState()
    )

    private val _userCustomPlaylistsState = MutableStateFlow<UiState<List<Playlist>>>(UiState.Idle)
    val userCustomPlaylistsState: StateFlow<UiState<List<Playlist>>> = _userCustomPlaylistsState

    private val _selectedArtistName = MutableStateFlow("")
    val selectedArtistName: StateFlow<String> = _selectedArtistName
    private val _selectedArtistMusicListState = MutableStateFlow<UiState<List<MusicInfo>>>(UiState.Idle)
    val selectedArtistMusicListState: StateFlow<UiState<List<MusicInfo>>> = _selectedArtistMusicListState

    private val _selectedAlbumName = MutableStateFlow("")
    val selectedAlbumName: StateFlow<String> = _selectedAlbumName
    private val _selectedAlbumMusicListState = MutableStateFlow<UiState<List<MusicInfo>>>(UiState.Idle)
    val selectedAlbumMusicListState: StateFlow<UiState<List<MusicInfo>>> = _selectedAlbumMusicListState

    private val _allMusicForAddPickerState = MutableStateFlow<UiState<List<MusicInfo>>>(UiState.Idle)
    val allMusicForAddPickerState: StateFlow<UiState<List<MusicInfo>>> = _allMusicForAddPickerState

    fun initializeDefaultPlaylists() {
        viewModelScope.launch {
            if (settingsRepository.getCurrentPlaylistId() == null) {
                managePlaylistUseCase.removePlaylist(name = getApplication<Application>().getString(R.string.default_playlist))
                val defaultId = managePlaylistUseCase.createPlaylist(name = getApplication<Application>().getString(R.string.default_playlist))
                settingsRepository.saveCurrentPlaylistId(defaultId)
            }

            if (settingsRepository.getLikedPlaylistId() == null) {
                managePlaylistUseCase.removePlaylist(name = getApplication<Application>().getString(R.string.heart))
                val likedId = managePlaylistUseCase.createPlaylist(name = getApplication<Application>().getString(R.string.heart))
                settingsRepository.saveLikedPlaylistId(likedId)
            }

            if (settingsRepository.getRecentPlaylistId() == null) {
                managePlaylistUseCase.removePlaylist(name = getApplication<Application>().getString(R.string.recently_played))
                val recentId = managePlaylistUseCase.createPlaylist(name = getApplication<Application>().getString(R.string.recently_played))
                settingsRepository.saveRecentPlaylistId(recentId)
            }
        }
    }

    init {
        initializeDefaultPlaylists()
        loadUserCustomPlaylists()
    }

    fun loadUserCustomPlaylists() {
        viewModelScope.launch {
            managePlaylistUseCase.getAllPlaylistsFlow()
                .catch {
                    _userCustomPlaylistsState.value = UiState.Error(it.message ?: "Failed to load playlists")
                }
                .collect { all ->
                    val playlists = all.filter { it.songCount > 0 }
                    _userCustomPlaylistsState.value = if (playlists.isEmpty()) {
                        UiState.Empty
                    } else {
                        UiState.Success(playlists)
                    }
                }
        }
    }

    private var currentPlaylistJob: Job? = null

    fun loadPlaylistById(playlistId: Long) {
        currentPlaylistJob?.cancel()
        _selectedPlaylistId.value = playlistId
        _selectedPlaylistState.value = UiState.Loading

        currentPlaylistJob = viewModelScope.launch {
            val meta = managePlaylistUseCase.getPlaylistMeta(playlistId)
            _selectedPlaylistMeta.value = meta
            _selectedPlaylistName.value = meta?.name ?: ""
            val currentId = settingsRepository.getCurrentPlaylistId()
            val likedId = settingsRepository.getLikedPlaylistId()
            val recentId = settingsRepository.getRecentPlaylistId()
            _isCustomPlaylist.value = playlistId != currentId && playlistId != likedId && playlistId != recentId

            managePlaylistUseCase.getMusicInfoInPlaylist(playlistId)
                .catch {
                    _selectedPlaylistState.value = UiState.Error(it.message ?: "Failed to load playlist")
                }
                .collect {
                    if (_selectedPlaylistId.value == playlistId) {
                        _selectedPlaylistState.value = if (it.isEmpty()) {
                            UiState.Empty
                        } else {
                            UiState.Success(it)
                        }
                    }
                }
        }
    }

    private fun refreshSelectedPlaylistMeta() {
        val id = _selectedPlaylistId.value ?: return
        viewModelScope.launch {
            val meta = managePlaylistUseCase.getPlaylistMeta(id)
            _selectedPlaylistMeta.value = meta
            meta?.name?.let { _selectedPlaylistName.value = it }
        }
    }

    fun createPlaylistAsync(name: String, onCreated: (Long) -> Unit) {
        viewModelScope.launch {
            val id = managePlaylistUseCase.createPlaylist(name)
            loadUserCustomPlaylists()
            onCreated(id)
        }
    }

    fun renamePlaylist(id: Long, newName: String) {
        viewModelScope.launch {
            managePlaylistUseCase.renamePlaylist(id, newName)
            if (_selectedPlaylistId.value == id) {
                _selectedPlaylistName.value = newName
                refreshSelectedPlaylistMeta()
            }
            loadUserCustomPlaylists()
        }
    }

    fun deletePlaylist(id: Long) {
        viewModelScope.launch {
            managePlaylistUseCase.removePlaylistById(id)
            _selectedPlaylistId.value = null
            _selectedPlaylistMeta.value = null
            _selectedPlaylistName.value = ""
            _selectedPlaylistState.value = UiState.Idle
            _isCustomPlaylist.value = false
            loadUserCustomPlaylists()
        }
    }

    fun removeItemFromPlaylist(musicId: Long, playlistId: Long) {
        viewModelScope.launch {
            managePlaylistUseCase.removeItemFromPlaylist(musicId, playlistId)
            if (playlistId == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun addItemToPlaylist(playlistId: Long, musicId: Long, musicPath: String) {
        viewModelScope.launch {
            managePlaylistUseCase.addToPlaylist(playlistId, musicId, musicPath)
            if (playlistId == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun addItemsToPlaylist(
        playlistId: Long,
        items: List<Pair<Long, String>>,
        onComplete: (() -> Unit)? = null
    ) {
        viewModelScope.launch {
            items.forEach { (musicId, musicPath) ->
                managePlaylistUseCase.addToPlaylist(playlistId, musicId, musicPath)
            }
            if (playlistId == _selectedPlaylistId.value) {
                refreshSelectedPlaylistMeta()
            }
            loadUserCustomPlaylists()
            onComplete?.invoke()
        }
    }

    fun loadAllMusicForAddPicker(onLoaded: ((List<MusicInfo>) -> Unit)? = null) {
        viewModelScope.launch {
            _allMusicForAddPickerState.value = UiState.Loading
            try {
                val allMusic = getAllMusicUseCase("title", "ASC")
                _allMusicForAddPickerState.value = if (allMusic.isEmpty()) {
                    UiState.Empty
                } else {
                    UiState.Success(allMusic)
                }
                onLoaded?.invoke(allMusic)
            } catch (e: Exception) {
                _allMusicForAddPickerState.value = UiState.Error(e.message ?: "Failed to load music")
            }
        }
    }

    fun reorderPlaylistItems(playlistId: Long, orderedMusicIds: List<Long>) {
        viewModelScope.launch {
            managePlaylistUseCase.reorderPlaylistItems(playlistId, orderedMusicIds)
            if (playlistId == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun recordPlaylistPlay(playlistId: Long) {
        viewModelScope.launch {
            managePlaylistUseCase.incrementPlaylistPlayCount(playlistId)
            managePlaylistUseCase.setPlaylistLastPlayedAt(playlistId, System.currentTimeMillis())
            if (playlistId == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
        }
    }

    fun updatePlaylistCover(id: Long, coverUri: String?) {
        viewModelScope.launch {
            managePlaylistUseCase.updatePlaylistCover(id, coverUri)
            if (id == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun updatePlaylistDescription(id: Long, description: String?) {
        viewModelScope.launch {
            managePlaylistUseCase.updatePlaylistDescription(id, description)
            if (id == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun setPlaylistPinned(id: Long, isPinned: Boolean) {
        viewModelScope.launch {
            managePlaylistUseCase.setPlaylistPinned(id, isPinned)
            if (id == _selectedPlaylistId.value) refreshSelectedPlaylistMeta()
            loadUserCustomPlaylists()
        }
    }

    fun getSelectedPlaylist(label: LabelName) {
        currentPlaylistJob?.cancel()
        _selectedPlaylistId.value = null
        _selectedPlaylistMeta.value = null
        _selectedPlaylistName.value = label.name
        _selectedPlaylistState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val musicList = musicLabelUseCase.getMusicListByLabel(label)
                _selectedPlaylistState.value = if (musicList.isEmpty()) {
                    UiState.Empty
                } else {
                    UiState.Success(musicList)
                }
            } catch (e: Exception) {
                _selectedPlaylistState.value = UiState.Error(e.message ?: "Failed to load playlist")
            }
        }
    }

    fun getSelectedPlaylist(label: String) {
        currentPlaylistJob?.cancel()
        _selectedPlaylistName.value = label
        _selectedPlaylistState.value = UiState.Loading
        viewModelScope.launch {
            val id = when (label) {
                getApplication<Application>().getString(R.string.default_list) -> settingsRepository.getCurrentPlaylistId()
                getApplication<Application>().getString(R.string.heart_list) -> settingsRepository.getLikedPlaylistId()
                getApplication<Application>().getString(R.string.recently_played) -> settingsRepository.getRecentPlaylistId()
                else -> null
            }
            if (id != null) {
                _selectedPlaylistId.value = id
                val currentId = settingsRepository.getCurrentPlaylistId()
                val likedId = settingsRepository.getLikedPlaylistId()
                val recentId = settingsRepository.getRecentPlaylistId()
                _isCustomPlaylist.value = id != currentId && id != likedId && id != recentId
                _selectedPlaylistMeta.value = managePlaylistUseCase.getPlaylistMeta(id)

                managePlaylistUseCase.getMusicInfoInPlaylist(id)
                    .catch {
                        _selectedPlaylistState.value = UiState.Error(it.message ?: "Failed to load playlist")
                    }
                    .collect {
                        if (_selectedPlaylistId.value == id) {
                            _selectedPlaylistState.value = if (it.isEmpty()) {
                                UiState.Empty
                            } else {
                                UiState.Success(it)
                            }
                        }
                    }
                return@launch
            }
            _selectedPlaylistId.value = null
            _selectedPlaylistMeta.value = null
            _isCustomPlaylist.value = false
            val labelEnum = LabelName.match(label)
            if (labelEnum != null) {
                try {
                    val musicList = musicLabelUseCase.getMusicListByLabel(labelEnum)
                    _selectedPlaylistState.value = if (musicList.isEmpty()) {
                        UiState.Empty
                    } else {
                        UiState.Success(musicList)
                    }
                } catch (e: Exception) {
                    _selectedPlaylistState.value = UiState.Error(e.message ?: "Failed to load playlist")
                }
            } else {
                _selectedPlaylistState.value = UiState.Empty
            }
        }
    }

    fun getSelectedArtistMusicList(artistName: String) {
        currentPlaylistJob?.cancel()
        _selectedArtistName.value = artistName
        _selectedArtistMusicListState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val musicList = getAllMusicUseCase.getMusicListByArtist(artistName)
                _selectedArtistMusicListState.value = if (musicList.isEmpty()) {
                    UiState.Empty
                } else {
                    UiState.Success(musicList)
                }
            } catch (e: Exception) {
                _selectedArtistMusicListState.value = UiState.Error(e.message ?: "Failed to load artist music")
            }
        }
    }

    fun getSelectedAlbumMusicList(albumName: String) {
        currentPlaylistJob?.cancel()
        _selectedAlbumName.value = albumName
        _selectedAlbumMusicListState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val musicList = getAllMusicUseCase.getMusicListByAlbum(albumName)
                _selectedAlbumMusicListState.value = if (musicList.isEmpty()) {
                    UiState.Empty
                } else {
                    UiState.Success(musicList)
                }
            } catch (e: Exception) {
                _selectedAlbumMusicListState.value = UiState.Error(e.message ?: "Failed to load album music")
            }
        }
    }
}